modules/network/src/main/java/org/apache/ignite/internal/network/netty/OutboundRecoveryHandler.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.network.netty;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import org.apache.ignite.internal.network.recovery.RecoveryDescriptor;
+import org.apache.ignite.network.OutNetworkObject;
+
+/** Outbound handler that adds outgoing message to the recovery descriptor. */
+public class OutboundRecoveryHandler extends ChannelOutboundHandlerAdapter {
+    /** Handler name. */
+    public static final String NAME = "outbound-recovery-handler";
+
+    /** Recovery descriptor. */
+    private final RecoveryDescriptor descriptor;
+
+    /**
+     * Constructor.
+     *
+     * @param descriptor Recovery descriptor.
+     */
+    public OutboundRecoveryHandler(RecoveryDescriptor descriptor) {
+        this.descriptor = descriptor;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        OutNetworkObject outNetworkObject = (OutNetworkObject) msg;
+
+        if (outNetworkObject.shouldBeSavedForRecovery()) {
+            descriptor.add(outNetworkObject);
```

Comment:
Another question - do we nullify serialized UOS fields? If we have already byte[] representations, then we don't need original objects. And, considering that we cache a bunch of messages, they **will** take a lot of space.

Reply:
Good point, I believe I tried nullifying this fields and there was some issues with that but I can't recall what exactly it was. Will look into that later

Reply:
Maybe they were final, or you tried to serialize them once again. It would be interesting to know